### PR TITLE
no-fork-no-reentry

### PR DESCRIPTION
**Pull request by Liaojinghui**
Alternative anti-reentry #2774 

This is a joint work from dusmart and me.

How to use it:

```C#
 public class Contract_Attribute : SmartContract
    {
        [NoReentry]
        public static bool test()
        {
            return true;
        }
    }
```

### CODE REVIEW

File: src/neo/SmartContract/ApplicationEngine.Runtime.cs [hunk 1]
```diff
@@ -354,5 +360,25 @@ protected internal void BurnGas(long gas)
...
+                if (_noReentryMethods[CurrentScriptHash].Contains(method + paramCount))
+                    throw new InvalidOperationException(
+                        $"Cannot Call Method {method} Of Contract {CurrentScriptHash} in an reentry manner.");
+            }
+            else
+            {
+                _noReentryMethods[CurrentScriptHash] = new List<string>();
+            }
+            _noReentryMethods[CurrentScriptHash].Add(method + paramCount);
+        }
```
Comment (dusmart) [inline src/neo/SmartContract/ApplicationEngine.Runtime.cs:382]: What if a method called 'Hello1' and parameter is count is '0' while another method is 'Hello' and parameter count is '10'?

**Comment by shargon** [comment]
Take a look to https://github.com/neo-project/neo/pull/2777

**Comment by Liaojinghui** [comment]
> Take a look to #2777

@shargon Two syscalls are not what i mean `make it easier for developers`, developers can do this even without syscall.

File: src/neo/SmartContract/ApplicationEngine.Runtime.cs (hunk 1 above)
Comment (Liaojinghui) [reply src/neo/SmartContract/ApplicationEngine.Runtime.cs:383]: @dusmart  addressed

**Pull request closed without merging**